### ISSUE TITLE
Error `$HOME must be set to run brew` when executing any Brew command

### ISSUE THREAD

**Issue by buresdv** [Homebrew/brew#15833]
### `brew doctor` output

```shell
Your system is ready to brew.
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.1.3-64-g3eb48a8
ORIGIN: https://github.com/Homebrew/brew
HEAD: 3eb48a828d44643b8d427f647227a277b1a4c1bf
Last commit: 27 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: ca97084daf9b750943e109c69772b2a4a62a33e9
Core tap last commit: 41 minutes ago
Core tap branch: master
Core tap JSON: 06 Aug 18:32 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Applications/Xcode-15.0.0-Beta.5.app/Contents/Developer/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.5-arm64
CLT: 14.3.1.0.1.1683849156
Xcode: 15.0 => /Applications/Xcode-15.0.0-Beta.5.app/Contents/Developer
Rosetta 2: false
```

### What were you trying to do (and why)?

I am the developer of [Cork](https://github.com/buresdv/Cork), a popular GUI for Homebrew. I am not sure if this a Homebrew or macOS problem, but contacting Apple is almost impossible, so I'm requesting help here to rule out this being a Homebrew problem, so I can focus my anger on Apple instead :)

Recently, users started reporting that they cannot execute any Brew command through Cork, such as commands for installing packages, adding Taps, etc.

This issue is experienced by every Cork user on the newest version of Homebrew.

The issue was traced back to every command given to Homebrew through the system's `Process()` class returning the error `$HOME must be set to run brew`. This issue is present even though Cork explicitly issues commands Homebrew using its full system path (`/opt/homebrew/bin/brew` on ARM and `/usr/local/bin/brew` on Intel).

See the following files for code reference:
- Implementation of `Process()` for issuing Homebrew commands and receiving their outputs: [Shell Interface.swift](https://github.com/buresdv/Cork/blob/73b19081768b403e271431a0fac19c5e4f128293/Cork/Logic/Shell/Shell%20Interface.swift#L44)
- Example use of this function to issue a Homebrew command: [Brew Interface.swift](https://github.com/buresdv/Cork/blob/73b19081768b403e271431a0fac19c5e4f128293/Cork/Logic/Shell/Brew%20Interface.swift#L20)
- Example of defining absolute path to the Homebrew executable: [App Constants.swift](https://github.com/buresdv/Cork/blob/73b19081768b403e271431a0fac19c5e4f128293/Cork/App%20Constants.swift#L13)

### What happened (include all command output)?

When issuing a Homebrew command through macOS's `Process()`, all Homebrew commands fail with the error `$HOME must be set to run brew` even when the Homebrew executable is executed with its explicit executable path (`/opt/homebrew/bin/brew` instead of just `brew`).

Example of output, see [Brew Interface.swift](https://github.com/buresdv/Cork/blob/a7310d99f5a76c04d41c0c93cf416e1cecd4e6e6/Cork/Logic/Shell/Brew%20Interface.swift#L34) for definition:
`TerminalOutput(standardOutput: "", standardError: "Error: $HOME must be set to run brew.\n")`

### What did you expect to happen?

The issued command returns a valid output, instead of an empty output and and the mentioned error.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
Example workflow: Loading outdated packages inside Cork.

1. When Cork starts, it uses Process() to issue the command "/opt/homebrew/bin/brew outdated" and awaits output.
2. Cork expects to receive a list of outdated packages, which it when displays in the UI

Instead, this happens:

1. Cork starts and uses Process() to issue the command "/opt/homebrew/bin/brew outdated" and awaits output.
2. Cork receives an output, STDOUT being empty, and STDERR including "$HOME must be set to run brew"

Again, I don't know if this is something that is happening because of Homebrew's newest version, or if macOS 13.5 is the culprit. But receiving a timely answer from Apple is impossible.
```

**Comment by Bo98** [comment]
We recently enforced this in https://github.com/Homebrew/brew/commit/4184546067cb33e8d4de3c0ce9b5e6f1ebf7daca, notably because brew never worked correctly without since `HOMEBREW_CACHE` would not be set correctly: https://github.com/Homebrew/brew/blob/3eb48a828d44643b8d427f647227a277b1a4c1bf/Library/Homebrew/brew.sh#L35-L45 (and this code has been there for 7 years).

Though I'm interested into hearing how you managed to get it to work before.

**Comment by MikeMcQuaid** [comment]
This is a Cork problem that you'll need to fix yourself, sorry. You can get the `$HOME` for a user and should set that in the environment accordingly. Otherwise, as @Bo98 has noted, several other parts of Homebrew will not work as expected.

**Comment by buresdv** [comment]
@Bo98 in the past, Homebrew didn't complain if I called the Homebrew executable by its full path even with an empty environment, so it until the `$HOME` thing was enforced, I didn't have any reasons to make any changes to it. I simply created a Swift `Process()`, set its launch path to the Brew executable, passed any required arguments to it and it worked just fine.

@MikeMcQuaid thank you for the info. I fixed the problem by getting the user's home directory and adding that as `HOME` to the `Process()`'s environment. And now it works as expected.
